### PR TITLE
Report root staleness checks

### PR DESCRIPTION
**Pull request by dimkouv**
Check in transmission protocol if report seq nums match the on-chain state. If not reject report transmission.

### CODE REVIEW

File: commit/plugin_functions.go [hunk 1]
```diff
@@ -585,6 +585,50 @@ func validateObservedGasPrices(gasPrices []cciptypes.GasPriceChain) error {
...
+		return true, nil
+	}
+
+	onchainSeqNums, err := reader.NextSeqNum(ctx, reportChains)
+	if err != nil {
+		return false, fmt.Errorf("get next sequence numbers: %w", err)
+	}
+
+	for i, onchainSeqNum := range onchainSeqNums {
+		if i >= len(reportChains) {
```
Comment (asoliman92) [inline commit/plugin_functions.go:612]: check before loop with `len(onchainSeqNums)`

File: commit/plugin_functions_test.go [hunk 2]
```diff
@@ -1180,3 +1180,80 @@ func Test_gasPricesConsensus(t *testing.T) {
...
+		},
+		{
+			name: "one root is stale",
+			reportSeqNums: []cciptypes.SeqNumChain{
+				cciptypes.NewSeqNumChain(10, 100),
+				cciptypes.NewSeqNumChain(20, 200),
+			},
+			onchainSeqNums: []cciptypes.SeqNum{99, 200}, // <- 200 is already on chain
+			expValid:       false,
+			expErr:         false,
```
Comment (asoliman92) [inline commit/plugin_functions_test.go:1210]: expErr should be true here, right?

File: commit/plugin_e2e_test.go [hunk 3]
```diff
@@ -324,6 +324,9 @@ func setupAllNodesReadAllChains(ctx context.Context, t *testing.T, lggr logger.L
 		n.ccipReader.On("NextSeqNum", ctx, []cciptypes.ChainSelector{chainA, chainB}).
 			Return([]cciptypes.SeqNum{10, 20}, nil)
 
+		// transmission phase root staleness check passes
+		n.ccipReader.On("NextSeqNum", ctx, []cciptypes.ChainSelector{chainB}).
+			Return([]cciptypes.SeqNum{20}, nil)
```
Comment (asoliman92) [inline commit/plugin_e2e_test.go:329]: Shouldn't it be  `21` to be valid?
```suggestion
			Return([]cciptypes.SeqNum{21}, nil)
```

Also any reason for not adding chainA?

```
		n.ccipReader.On("NextSeqNum", ctx, []cciptypes.ChainSelector{chainA}).
			Return([]cciptypes.SeqNum{11}, nil)
```

File: commit/plugin_functions.go [hunk 4]
```diff
@@ -585,6 +585,50 @@ func validateObservedGasPrices(gasPrices []cciptypes.GasPriceChain) error {
...
+	for _, mr := range report.MerkleRoots {
+		reportChains = append(reportChains, mr.ChainSel)
+		reportMinSeqNums[mr.ChainSel] = mr.SeqNumsRange.Start()
+	}
+
+	if len(reportChains) == 0 {
+		return true, nil
+	}
+
+	onchainSeqNums, err := reader.NextSeqNum(ctx, reportChains)
```
Comment (makramkd) [inline commit/plugin_functions.go:606]: Lets disambiguate and make it clear which sequence numbers we're talking about


```suggestion
	onchainNextExpectedSeqNums, err := reader.NextSeqNum(ctx, reportChains)
```

File: commit/plugin_functions.go [hunk 5]
```diff
@@ -585,6 +585,50 @@ func validateObservedGasPrices(gasPrices []cciptypes.GasPriceChain) error {
 	return nil
 }
 
+// validateMerkleRootsState merkle roots seq nums validation by comparing with on-chain state.
+func validateMerkleRootsState(
+	ctx context.Context,
+	lggr logger.Logger,
+	report cciptypes.CommitPluginReport,
+	reader cciptypes.CCIPReader,
+) (bool, error) {
```
Comment (makramkd) [inline commit/plugin_functions.go:594]: Will this func gracefully fail (i.e return an error) if the node doesn't support the dest?

**Review by makramkd — COMMENTED** [review]
Minor comments

File: commit/plugin_e2e_test.go (hunk 3 above)
Comment (dimkouv) [reply commit/plugin_e2e_test.go:329]: if 21 was already committed we wouldn't be able to post 21

Comment (dimkouv) [reply commit/plugin_e2e_test.go:329]: the report of this test does not include chainA

File: commit/plugin_functions.go (hunk 4 above)
Comment (dimkouv) [reply commit/plugin_functions.go:606]: I'll rename the interface method, contract returns the last committed seq num. And rename var to `onchainCommittedSeqNums`

File: commit/plugin_functions.go (hunk 5 above)
Comment (dimkouv) [reply commit/plugin_functions.go:594]: if the node does not support the dest this is not called at all

Comment (dimkouv) [reply commit/plugin_functions.go:594]: caller provides a reader

File: commit/plugin_e2e_test.go (hunk 3 above)
Comment (dimkouv) [reply commit/plugin_e2e_test.go:329]: we have to rename the interface method name to ``LastCommittedSeqNums` not `NextSeqNums`. I'll do it as a followup

File: commit/plugin_functions.go (hunk 4 above)
Comment (dimkouv) [reply commit/plugin_functions.go]: https://github.com/smartcontractkit/chainlink-common/pull/627

**Review by asoliman92 — APPROVED** [review]
LGTM

**Review by makramkd — APPROVED** [review]

**Pull request merged**